### PR TITLE
Feature auto margin

### PR DESCRIPTION
**Pull request by woehrl01**
Even so I know there are some opinions against ```margin: 0 auto``` it's still part of the spec: https://www.w3.org/TR/css-flexbox-1/#auto-margins and pretty usefull if you have to position via ```justify-content```. 

This PR adds an implementation for that.

It adds an additonal ```YGUnitAuto``` and margins got ```YGNodeStyleSetMarginAuto``` functions as well.

### CODE REVIEW

**Comment by emilsjolander** [comment]
@woehrl01 We don't aim to implement the full css spec. Could you add some examples of layouts which are made significantly easier / possible with margin auto support compared to implementing those layouts today?

**Comment by woehrl01** [comment]
@emilsjolander I didn't came across any layout yet, which wasn't able to implement with yoga today. But I had to add spacer views. I want to get rid of spacer views as much as I can and the last "thing" can be only solved with auto margins. 

I'm totally aware that there is no need for the full css spec, but as auto margins are explicitly mentioned and described in flex-box I would still account it to the flex-box model layouting logic.

**Comment by emilsjolander** [comment]
I would count needing spacer views as a limitation. Could you give some examples of where margin auto would remove the need for spacer views?

**Comment by woehrl01** [comment]
Sure, see this simple [JSFiddle](https://jsfiddle.net/uojyuspt/).

It's pretty much any layout were you want to have elements on one side and some additonal on the other side e.g for menus.

I also use this e.g. for centered notification messages where I have additional text at the bottom (can be boiled down to the layout above but in column direction)

**Comment by emilsjolander** [comment]
Thanks, that helps!

File: gentest/fixtures/YGMarginTest.html [hunk 1]
```diff
@@ -15,27 +15,73 @@
 </div>
 
 <div id="margin_and_flex_row" style="width: 100px; height: 100px; flex-direction: row;">
-  <div style="margin-start: 10px; margin-end; 10px; flex-grow: 1;"></div>
```
Comment (emilsjolander) [inline gentest/fixtures/YGMarginTest.html]: This should be a fix in a seperate pull request.

File: gentest/gentest-cpp.js [hunk 2]
```diff
@@ -15,6 +15,8 @@ function toValueCpp(value) {
 function toFunctionName(value) {
   if (value.indexOf('%') >= 0){
     return 'Percent';
+  }else if(value.indexOf('Auto') >= 0){
```
Comment (emilsjolander) [inline gentest/gentest-cpp.js]: space before `else`

File: gentest/gentest-cpp.js [hunk 3]
```diff
@@ -189,7 +193,13 @@ CPPEmitter.prototype = Object.create(Emitter.prototype, {
   }},
 
   YGNodeStyleSetMargin:{value:function(nodeName, edge, value) {
-    this.push('YGNodeStyleSetMargin' + toFunctionName(value) + '(' + nodeName + ', ' + edge + ', ' + toValueCpp(value) + ');');
+    var valueStr = toValueCpp(value);
+    if(valueStr != 'YGAuto'){
```
Comment (emilsjolander) [inline gentest/gentest-cpp.js]: space before `(` and after `)`

File: gentest/gentest-cpp.js [hunk 4]
```diff
@@ -189,7 +193,13 @@ CPPEmitter.prototype = Object.create(Emitter.prototype, {
...
+      valueStr = ', ' + valueStr;
+    }else{
```
Comment (emilsjolander) [inline gentest/gentest-cpp.js]: space before and after `else`

**Comment by emilsjolander** [comment]
We currently use `YGUndefined` to mean `auto` for `width`, `height`, and `flex-basis`. I'm concerned with this being confusing after this PR introduced an explicit `auto` value for margin. If we introduce a specific `auto` value then I think this should be respected for other properties as well.

**Comment by woehrl01** [comment]
@emilsjolander I thought about this too, should I add/change those? I would handle undefined the same like auto would on those values, so this wouldn't brake existing users.

**Comment by emilsjolander** [comment]
@woehrl01 I'm not sure about the best way to handle it, feel free to experiment. But yes, whatever we do we should not break existing usages.

**Comment by woehrl01** [comment]
@emilsjolander I changed ```width```, ```height``` and ```flex-basis``` to have ```auto``` as default value. undefined on the setter is silently converted to auto.

**Comment by emilsjolander** [comment]
@woehrl01 looks pretty good. I'll get back to you later this week regarding this.

File: gentest/fixtures/YGMarginTest.html [hunk 5]
```diff
@@ -39,3 +39,54 @@
...
+  <div style="width: 50px; height: 50px;margin-top:auto;margin-bottom:auto"></div>
+  <div style="width: 50px; height: 50px;"></div>
+</div>
+
+<div id="margin_auto_bottom_and_top_justify_center" style="width: 200px; height: 200px; justify-content: center;">
+  <div style="width: 50px; height: 50px;margin-top:auto;margin-bottom:auto"></div>
+  <div style="width: 50px; height: 50px;"></div>
+</div>
+
+<div id="margin_auto_top_double" style="width: 200px; height: 200px; align-items: center;">
```
Comment (emilsjolander) [inline gentest/fixtures/YGMarginTest.html]: add a similar test for row direction. Also please change name to something like `margin_auto_multiple_children` for a `_row` and `_column` suffix.

**Comment by woehrl01** [comment]
I added most of the language bindings except objc and javascript. As I don't know objc good enough and for javascript I think it would be good to extend ```entry-common``` to also support ```"auto"``` as an additional short cut (the tests are generated for this) but I'm not familiar enough with this change.

**Comment by emilsjolander** [comment]
@woehrl01 that's fine! I'll fix up javascript and objc when importing 👍 Thanks for this awesome addition 💯

**Comment by emilsjolander** [comment]
@woehrl01 could you rebase this? Github says there are no conflicts but i'm having some issues importing it.

**Comment by woehrl01** [comment]
@emilsjolander merged the master for this and the other branches you didn't imported yet.

**Comment by emilsjolander** [comment]
@woehrl01 Could you rebase this on top of master instead of merging master into this branch?

**Comment by woehrl01** [comment]
@emilsjolander I'll try my best, I'm more of a mercurial than a git user. 😊

**Comment by emilsjolander** [comment]
@woehrl01 lol me to. Another option is to squash the commit and the push it, that should work as well when I import.

**Comment by woehrl01** [comment]
@emilsjolander let me know if it works, otherwise I'll squash them all. 💪

**Comment by emilsjolander** [comment]
@woehrl01 it didn't work :/ can you try squashing?

**Comment by woehrl01** [comment]
@emilsjolander try this, please 🤘

**Comment by emilsjolander** [comment]
@woehrl01 that worked! Thanks! sorry for the hassle.

**Comment by woehrl01** [comment]
@emilsjolander 👏 awesome! No worries, thanks to this I learn new things about git today 👍

**Comment by emilsjolander** [comment]
I'm still working on this. It breaks some tests internally and I have a couple other things that are higher priority. hang tight 👍

**Comment by woehrl01** [comment]
@emilsjolander If there is anything were I can help you out. Let me know it! No rush 👍

**Comment by emilsjolander** [comment]
@woehrl01 will do!

**Comment by woehrl01** [comment]
@emilsjolander I did a rebase and fixed the auto margin as some of the behaviour this implicitly used has changed, now its explicit.

**Pull request closed without merging**